### PR TITLE
Adding method to check namelist in any order, tested with NAG Fortran.

### PR DESCRIPTION
**Pull request by daveh150**
## PR checklist
- [x] Short (1 sentence) summary of your PR: 
    Check namelist in any order to support NAG compiler.
- [x] Developer(s): 
    David Hebert
- [x] Suggest PR reviewers from list in the column to the right.
- [x] Please copy the PR test results link or provide a summary of testing completed below.
ran quicksuite, modified for tests to run with 1 PE. Compiled main branch with gnu for baseline. 
Then compiled check_namelist2  branch with gnu and nag compilers. Results show all tests past. 
Also added errors to namelists. Code crashes with message on what line has error.:

(abort_ice)ABORTED:
 (abort_ice) called from ice_init.F90
 (abort_ice) line number  644
 (abort_ice) error = (input_data)ERROR: grid_nml reading     test_namelist = .true.

- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [x] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [x] No 
- [x] Please provide any additional information or relevant details below:

Re-opening #788. Implemented method to rewind and search ice_in for namelist, then read namelist and check for errors. This is done through addition of subroutine goto_nml_group in ice_namelist_mod.F90. I added new file because of circular dependencies between ice_in.F90 and ice_domain.F90. 

I have a trial NAG FORTRAN compiler license for a few more days, this has been tested with serial NAG and GNU compiler.  I inserted error in namelists and run crashes with message like:

 (abort_ice)ABORTED:
 (abort_ice) called from ice_init.F90
 (abort_ice) line number  644
 (abort_ice) error = (input_data)ERROR: grid_nml reading     test_namelist = .true.

I also reordered the namelists, putting grid_nml before setup_nml, and both compilers worked.  

I admittedly do not have mpi fortran compiled with NAG, but since this is only a namelist check I am hoping mpi vs. serial won't matter. If someone has access to machine with NAG FORTRAN MPI to do a more thorough test that would be helpful.  

Disclosure, after much googling for namelist, method is based on this link:
https://stackoverflow.com/questions/57553995/how-to-handle-an-optional-group-in-a-fortran-namelist

### CODE REVIEW

**Comment by apcraig** [comment]
Thanks @daveh150 for continuing to pursue this!  This looks great assuming it's robust.  I can run some tests on more compilers/platforms (although I don't have access to nag at the moment either).  One comment, rather than create a new file for the new subroutine, how about dropping it into cicecore/shared/ice_fileunits.F90.  I think it's both a reasonable place and shouldn't have any problems with circular logic.  Thoughts?

**Comment by daveh150** [comment]
Thanks @apcraig! Sure, moving the goto_nml_group to ice_fileunits.F90 seems the best location without adding more files. I'll make that move now.

**Comment by daveh150** [comment]
I moved the goto_ nml subroutine to ice_fileunits.F90. If there are any other thoughts or suggestions please let me know.

**Comment by apcraig** [comment]
FYI, I plan to defer this PR until next week, after the release.  I want to be able to adequately test on several platforms and make sure everything is robust.  We've had issue in the past, although I think this solution should be fine.  Just don't want to introduce unnecessary risk this late.  Thanks.

File: cicecore/cicedyn/analysis/ice_history_bgc.F90 [hunk 1]
```diff
@@ -305,24 +309,41 @@ subroutine init_hist_bgc_2D
       !-----------------------------------------------------------------
 
       if (my_task == master_task) then
-         write(nu_diag,*) subname,' Reading icefields_bgc_nml'
+        nml_name = 'icefields_bgc_nml'
```
Comment (apcraig) [inline cicecore/cicedyn/analysis/ice_history_bgc.F90:312]: Looks like this line needs an extra space for alignment.

File: cicecore/cicedyn/analysis/ice_history_fsd.F90 [hunk 2]
```diff
@@ -76,10 +76,13 @@ subroutine init_hist_fsd_2D
       use ice_calendar, only: nstreams
       use ice_communicate, only: my_task, master_task
       use ice_history_shared, only: tstr2D, tcstr, define_hist_field
+      use ice_fileunits, only: goto_nml
 
       integer (kind=int_kind) :: ns
       integer (kind=int_kind) :: nml_error ! namelist i/o error flag
       logical (kind=log_kind) :: tr_fsd, wave_spec
+      character (len=char_len_long) :: tmpstr2  ! test namelist
```
Comment (apcraig) [inline cicecore/cicedyn/analysis/ice_history_fsd.F90:84]: alignment?  see same code in other subroutines?

```
      character(len=char_len_long) :: tmpstr2 ! for namelist check
      character(len=char_len)      :: nml_name ! for namelist check
```

File: cicecore/cicedyn/general/ice_init.F90 [hunk 3]
```diff
@@ -592,141 +596,244 @@ subroutine input_data
...
          do while (nml_error > 0)
             read(nu_nml, nml=forcing_nml,iostat=nml_error)
+            ! check if error
+            if (nml_error /= 0) then
+               ! backspace and re-read erroneous line
+               backspace(nu_nml)
+               read(nu_nml,fmt='(A)') tmpstr2
+               call abort_ice(subname//'ERROR: '// trim(nml_name)//' reading '// &
+                    trim(tmpstr2), &
+                    file=__FILE__, line=__LINE__)
```
Comment (apcraig) [inline cicecore/cicedyn/general/ice_init.F90:832]: same, check other instances too

File: cicecore/cicedyn/infrastructure/ice_domain.F90 [hunk 4]
```diff
@@ -167,26 +170,41 @@ subroutine init_domain_blocks
...
+      write(nu_diag,*) subname,' Reading ', trim(nml_name)
+      
       call get_fileunit(nu_nml)
       open (nu_nml, file=trim(nml_filename), status='old',iostat=nml_error)
       if (nml_error /= 0) then
          call abort_ice(subname//'ERROR: domain_nml open file '// &
-            trim(nml_filename), &
-            file=__FILE__, line=__LINE__)
+              trim(nml_filename), &
+              file=__FILE__, line=__LINE__)
```
Comment (apcraig) [inline cicecore/cicedyn/infrastructure/ice_domain.F90:181]: also, alignment across new abort_ice calls seems to vary.  could it be more consistent?

**Comment by apcraig** [comment]
Testing seems to be good.  I think we can merge this.  Maybe one more quick pass looking at the indentation things noted above, then we can merge.  Thanks.

**Comment by daveh150** [comment]
Great, thanks Tony! I’ll clean up the indentation now.

File: cicecore/cicedyn/general/ice_init.F90 [hunk 5]
```diff
@@ -592,141 +596,244 @@ subroutine input_data
...
             read(nu_nml, nml=snow_nml,iostat=nml_error)
+            ! check if error
+            if (nml_error /= 0) then
+               ! backspace and re-read erroneous line
+               backspace(nu_nml)
+               read(nu_nml,fmt='(A)') tmpstr2
+
+               call abort_ice(subname//'ERROR: '//trim(nml_name)//' reading '// &
+                    trim(tmpstr2), &
+                    file=__FILE__, line=__LINE__)
```
[the remark this answers is not in the document]
Comment (daveh150) [reply cicecore/cicedyn/general/ice_init.F90:799]: I found the line to be too long with the entire concatenated string on a single line. I prefer to leave it as it, but if others find it too confusing I can put it all on one line.

[the remark this answers is not in the document]
Comment (apcraig) [reply cicecore/cicedyn/general/ice_init.F90:799]: I might combine the last two lines, tmpstr2 and file/line, leaving first line alone?

[the remark this answers is not in the document]
Comment (daveh150) [reply cicecore/cicedyn/general/ice_init.F90]: Okay, I see now. I combined the last to to read trim (tmpstr2), file=...

I think I have all other intents/spacings updated. Let me know if I missed anything. Thanks!

**Review by apcraig — APPROVED** [review]

**Pull request merged**